shotover/src/server.rs

```diff
@@ -233,50 +345,134 @@ impl<C: CodecBuilder + 'static> TcpCodecListener<C> {
                         // Only prune the list every so often
                         // theres no point in doing it every iteration because most likely none of the handles will have completed
                         if self.connection_count % 1000 == 0{
                             self.connection_handles.retain(|x| !x.is_finished());
                         }
-                        Ok::<(), anyhow::Error>(())
+                        Ok::<bool, anyhow::Error>(false)
                     },
                     // Hot reload request handling
                     hot_reload_request = self.hot_reload_rx.recv() => {
                         if let Some(request) = hot_reload_request{
                             self.handle_hot_reload_request(request).await;
-                            // Wait forever once the FD has been sent. This prevents the loop from continuing
-                            // and attempting to recreate the listener.
-                            // This is fine, since the TcpCodecListener has no more work to do once it has handed off its listener.
-                            // Unfortunately, simply returning from `run` would not work as that would cause shotover to shutdown since there are no more sources running.
-                            futures::future::pending().await
+                            // After handing off the socket FD, mark it as handed off and clear the listener
+                            // This prevents attempting to recreate the listener on the same address
+                            self.socket_handed_off = true;
+                            self.listener = None;
                         }
-                        Ok::<(), anyhow::Error>(())
-                    }
-                }
+                        Ok::<bool, anyhow::Error>(false)
+                    },
+                    // Gradual shutdown request handling
+                    gradual_shutdown_request = self.gradual_shutdown_rx.recv() => {
+                        if let Some(request) = gradual_shutdown_request{
+                            info!("[{}] Received gradual shutdown request", self.source_name);
+                            self.handle_gradual_shutdown_request(request).await;
+                            info!("[{}] Gradual shutdown initiated", self.source_name);
+                        } else {
+                            // Channel was closed, but this is expected during normal shutdown
+                            // Just continue the loop - shotover shutdown will be triggered via the shutdown signal
+                            debug!("[{}] Gradual shutdown channel closed", self.source_name);
+                        }
+                        Ok::<bool, anyhow::Error>(false)
+                    },
+                    // Gradual shutdown draining - runs every 10 seconds when in shutdown mode
+                    _ = self.drain_interval.tick(), if self.gradual_shutdown_in_progress => {
```

Comment:
Do we need this `_ = self.drain_interval.tick(), if self.gradual_shutdown_in_progress => {` state?
Cant we just run the shutdown logic immediately in `gradual_shutdown_request = self.gradual_shutdown_rx.recv() => { .. }`?